infra/presubmit.py

```diff
@@ -0,0 +1,173 @@
+  script_path = os.path.realpath(__file__)
+  return os.path.abspath(os.path.join(os.path.dirname(script_path), '..'))
+
+
+def is_project_file(actual_path, expected_filename):
+  if os.path.basename(actual_path) != expected_filename:
+    return False
+
+  if os.path.basename(
+      os.path.dirname(os.path.dirname(actual_path))) != 'projects':
+    return False
+
+  return os.path.exists(actual_path)
+
+
+def check_lib_fuzzing_engine(build_sh_file):
+  if not is_project_file(build_sh_file, 'build.sh'):
+    return True
+
+  with open(build_sh_file) as build_sh:
+    build_sh_lines = build_sh.readlines()
+  for line_num, line in enumerate(build_sh_lines):
+    uncommented_code = line.split('#')[0]
+    if '-lFuzzingEngine' in uncommented_code:
+      print('''Error: build.sh contains -lFuzzingEngine on line: {0}.
+Please use $LIB_FUZZING_ENGINE.'''.format(line_num))
+      return False
+  return True
+
+
+class ProjectYamlChecker:
+  def __init__(self, project_yaml_filename):
+    self.project_yaml_filename = project_yaml_filename
+    with open(project_yaml_filename) as file_handle:
+      self.project_yaml = yaml.safe_load(file_handle)
+
+    self.success = True
+
+    self.checks = [self.check_project_yaml_constants,
+                   self.check_required_sections,
+                   self.check_valid_section_names,
+                   self.check_valid_emails]
+
+  def do_checks(self):
+    """Do all project.yaml checks. Return True if they pass."""
+    if self.is_disabled():
+      return True
+    for check_function in self.checks:
+      check_function()
+    return self.success
+
+  def is_disabled(self):
+    """Is this project disabled."""
+    return self.project_yaml.get('disabled', False)
+
+  def print_error_message(self, message, *args):
+    """Print an error message and set self.success to False."""
+    self.success = False
+    message = message % args
+    print('Error in %s: %s' % (self.project_yaml_filename, message))
+
+  def check_project_yaml_constants(self):
+    """Check that certain sections only have certain constant values."""
+    sections_and_constants = {
+        'sanitizers': ['address', 'none', 'memory', 'address'],
+        'architectures': ['i386', 'x86_64'],
+        'engines': ['afl', 'libfuzzer', 'honggfuzz']
+    }
+    success = True
+    for section, constants in sections_and_constants.items():
+      if section not in self.project_yaml:
+        continue
+      section_contents = self.project_yaml[section]
+      for constant in section_contents:
+        if constant not in section_contents:
+          self.print_error_message('%s not one of %s', constant, constants)
+
+  def check_valid_section_names(self):
+    """Check that only valid sections are included."""
+    # Note, this list must be updated when we allow new sections.
+    valid_section_names = [
+        'homepage', 'primary_contact', 'auto_ccs', 'sanitizers',
+        'architectures', 'disabled'
```

Comment:
also need `'view_restrictions'`, `'coverage_extra_args'`, and `'vendor_ccs'`. Maybe there area others too @Dor1s @inferno-chromium  ?

Reply:
Done.